src/box/box.cc

```diff
@@ -793,6 +799,172 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
+	int64_t lsr_id;
+};
+
+static void
+wal_mem_init(void)
+{
+	mempool_create(&wal_row_pool, &cord()->slabc,
+		       sizeof(struct wal_row));
+	lsregion_create(&wal_row_lsr, &runtime);
+}
+
+static void
+wal_mem_destroy(void)
+{
+	lsregion_destroy(&wal_row_lsr);
+	mempool_destroy(&wal_row_pool);
+}
+
+/*
+ * Callback to stash row upon receipt, used to recover mixed transactions.
+ */
+static struct wal_row *
+wal_stream_save_row(struct xrow_header *row)
+{
+	struct wal_row *new_row =
+	(struct wal_row *)mempool_alloc(&wal_row_pool);
+	if (new_row == NULL) {
+		tnt_raise(OutOfMemory, sizeof(*new_row), "mempool_alloc",
+			  "xrow");
+	}
+	memcpy(new_row, (struct wal_row *)row, sizeof(struct wal_row));
+	return new_row;
+}
+
+/*
+ * Callback to stash row bodies upon receipt, used to recover mixed
+ * transactions.
+ */
+static void
+wal_stream_save_body(struct wal_row *row)
+{
+	static int64_t id = 0;
+	assert(row->row.bodycnt <= 1);
+	if (row->row.bodycnt == 1) {
+		size_t len = row->row.body[0].iov_len;
+		char *new_base = (char *)lsregion_alloc(&wal_row_lsr, len,
+							++id);
+		if (new_base == NULL)
+			tnt_raise(OutOfMemory, len, "lsregion_alloc",
+				  "xrow body");
+		row->lsr_id = id;
+		memcpy(new_base, row->row.body[0].iov_base, len);
+		/* Adjust row body pointers. */
+		row->row.body[0].iov_base = new_base;
+	}
+}
+
+static struct rlist nodes_rows[VCLOCK_MAX];
+
+/*
+ * When restoring the log is read row-by-row. However, it is necessary
+ * to store the rows for correct further recovery. For example, with
+ * mixed transactions. The following function saves the newly arrived
+ * row to the rlist, which is stored in array is nodes_rows in cell is
+ * replica_id. As soon as a row arrives with the is_commit flag set, the
+ * corresponding transaction tries to apply. If an error occurs the
+ * transaction is rolled back. And we continue to work with the next
+ * transaction.
+ */
+static int
+wal_stream_apply_mixed_dml_row(struct wal_stream *stream,
+			       struct xrow_header *row)
+{
+	int rc = 0;
+	uint32_t id = row->replica_id;
+	/* The rlist can be created if the cell of array is empty */
+	if (!nodes_rows[id].prev && !nodes_rows[id].next) {
+		rlist_create(&nodes_rows[id]);
+	}
+
+	/* When a new row arrives, it has the same address. Because the
+	 * row is on the stack. Therefore, it needs to allocate memory on
+	 * the heap. We do with row->body[0].iov_base analogically.
+	 */
+	struct wal_row *save_row = wal_stream_save_row(row);
+	wal_stream_save_body(save_row);
+
+	rlist_add_tail_entry(&nodes_rows[id], save_row, in_row_list);
+
+	if (!row->is_commit) {
+		return 0;
+	} else {
+		struct wal_row *item, *next_row;
+		rlist_foreach_entry_safe(item, &nodes_rows[id],
+					 in_row_list, next_row) {
+			rlist_del_entry(item, in_row_list);
+			if (rc == 0) {
+				rc = wal_stream_apply_dml_row(stream,
+				     (struct xrow_header *)item);
+			}
+			/* deallocating the memory */
+			lsregion_gc(&wal_row_lsr, item->lsr_id);
```

Comment:
This is not right.
Imagine you have an xlog
1.`replica 1 lsn 1 begin`
2.`replica2 lsn 1 begin`
3.`replica1 lsn2 commit`
4.`replica2 lsn2 commit`.
The first row will receive lsr_id 1, the second one lsr_id 2 and so on.

By the time you read row 3, which is commit one, you call `lsregion_gc` on each row of transaction. It will be `lsregion_gc(..., lsr_id = 1)` and `lsregion_gc(..., lsr_id = 3)`. So you'll free row 2 by accident. And it is still needed.

You need to track the smallest lsr_id still needed, and call lsregion_gc() on that lsr_id, not on the lsr_id of the applied row.

Reply:
Good job!